rfcs/0000-user-defined-aggregate-functions.md

```diff
@@ -0,0 +1,269 @@
+---
+feature: user_defined_aggregate_functions
+authors:
+  - "Runji Wang"
+start_date: "2023/08/02"
+---
+
+# User Defined Aggregate Functions (UDAF)
+
+## Summary
+
+This RFC proposes the user interface and internal design of user defined aggregate functions (UDAF).
+
+## User Interfaces
+
+This section describes how users create UDAFs in Python, Java, and SQL through examples.
+
+### Python API
+
+Similar to scalar functions and table functions, we provide an `@udaf` decorator to define aggregate functions. The difference is that below the decorator, we define a class instead of a function. The class has an associated **accumulator** type, which is the intermediate state of the aggregate function.
+
+```python
+from risingwave.udf import udaf
+
+# The accumulator (intermediate state) can be arbitrary class.
+# It will be serialized into bytes before sending to kernel,
+# and deserialized from bytes after receiving from kernel.
+class State:
+    sum: int
+    count: int
+
+# The aggregate function is defined as a class.
+# Specify the schema of the aggregate function in the `udaf` decorator.
+@udaf(input_types=['BIGINT', 'INT'], result_type='BIGINT')
+class WeightedAvg:
+    # Create an empty accumulator.
+    def create_accumulator(self) -> State:
+        accumulator = State()
+        accumulator.sum = 0
+        accumulator.count = 0
```

Comment:
In Python we can also do the following so that no need to manually initialize each field (if we meant to use this as a demo in document)

```py
class State:
    sum: int = 0
    count: int = 0
```